Add no_std compatibility

SzymonKubica opened the pull request:
### What was added?

Added support for rbpf to run on `no_std` environments. 

### Why is it needed?

Using rbpf on targets without `std` support increases compatibility of the crate. 
I've been using a modified version of rbpf to run eBPF programs on microcontrollers (see [here(https://github.com/SzymonKubica/mibpf))

### Testing

Currently the code builds successfuly with both configuration options enabled, i.e. all build configurations succeed:
```
cargo build
cargo build --features cranelift
cargo build --no-default-features
cargo build --features cranelift --no-default-features
```

All tests (except for the ones using the `print` - only available in std) pass in both configuration options.

### TODO items for the  PR

- [x] test `no_std` on an example embedded device

### Notes

The structure of `std` / `no_std` feature-gating is inspired by `serde`, inside of `lib.rs` a public module: `lib` is defined and it 
contains reexports of all required items from std/alloc/core. It is then used throughout the codebase by importing all items 
declared inside of it: 
```
use crate::lib::*;
```
This behaves similarly as if we were using the standard library as we don't have to import things like `String` or `Vec` manually.

qmonnet commented on src/without_std.rs:
```diff
@@ -3,7 +3,6 @@ pub mod without_std {
     use alloc::string::String;
     pub use core::u32;
     pub use core::u64;
-
```
Why removing this line?

qmonnet commented on src/lib.rs:
```diff
@@ -64,7 +64,7 @@ mod stdlib {
 use byteorder::{ByteOrder, LittleEndian};
 use stdlib::collections::HashMap;
-use stdlib::collections::Vec;
+use stdlib::vec::Vec;
```
If such changes are necessary to “make it build successfully”, they should likely be fold into the previous commit. Again, more descriptive commit logs to understand what happens would be nice :)

qmonnet commented on examples/rbpf_plugin.rs:
```diff
@@ -42,7 +42,7 @@ fn main() {
                     return;
                 }
-                #[cfg(jit)] {
+                #[cfg(feature = "jit")] {
```
Please squash these with the first commit.

SzymonKubica replied:
Ok that's done, would you say I should keep all changes that I add with my PR within a single commit so that it is easier to review? It is currently done like this

qmonnet replied:
No, please don't do a single commit for all the PR! Some of the steps are logically separated, like adding the `jit` feature flag and then handling the changes for no_std support.

What I would like is a set of logical, but clean commits. In a sense that each commit should contain a logical change, _and_ should be self-contained. There shouldn't be later commit with fixes for the first commits; instead, all “fixes” should be squashed into the commits they improve, to make sure each commit works on its own, and doesn't break the build.

Does it make sense?

SzymonKubica replied:
Yes that makes sense I will do my best to follow this practice. Thank you for the feedback, that's really helpful and I agree that it would make it much easier to review and then go through later on.

qmonnet replied:
You're welcome. Kernel docs give more guidance about this, if you wanted more:
- https://docs.kernel.org/process/submitting-patches.html#separate-your-changes
- https://docs.kernel.org/process/5.Posting.html#patch-preparation

The objective is to be able to review logical changes independently, and efficiently. One motivation to preserve the builds through each commit is to be able to `git bisect` efficiently, if needed.

I'll review the rest of your changes soon. Thanks a lot!

SzymonKubica replied:
Ok thank you for the additional pointers, I've read through those, that's really useful.
Now that I think about this, the jit flag complicates things quite a bit and I think I can get around that using combinations of cfg not(windows) and feature = "std" (see replies above). In this case I think I will open up a new PR and apply the current changes there in a clean way in a single commit. The thing with adding no_std is that I can't really do it half way through so (it needs to be 'atomic') otherwise it won't compile without std. In which case I think it is a single logical change and could be put into a single commit. 

Thank you for all of your comments and suggestions in this thread. Is it ok if I open the new clean PR and have the final version of changes there without all of this intermediate changes that happened here? Or should I continue here so that our conversation gets preserved?

qmonnet replied:
You're free to open a new PR, but I don't really see the point to be honest, it's perfectly fine to update and force-push here.

I'd still prefer having several, smaller commits. For example, adding the `std` feature to the Cargo.toml and using it to gate the related code in addition to `windows` is a logical change and doesn't break the build, even if the `std` feature has not been implemented yet. I'm pretty sure some other parts can be their own commits, too. Don't lose too much time on that, but generally speaking, just keep in mind that smaller commits are easier to review and understand in the future.

SzymonKubica replied:
Ok sounds good, thank you for your answer. I do agree that opening up a new RP could potentially create additional mess. I will force-push the clean changes to this one then and try to break them up into separate logical commits

qmonnet commented on Cargo.toml:
```diff
@@ -43,7 +43,8 @@ json = "0.11"
 [features]
-default = ["std"]
+default = ["std", "jit"]
+jit = []
```
This would also need to go into the first commit.

It's still subject to the caveats regarding support for Windows.

SzymonKubica commented:
Hi, 
Thank you for reviewing the PR and listing the required changes. 
I will start going through these and clarify in the comments above if anything turns out to be unclear. 

No worries about the delay, my goal with this PR was to contribute back a small subset of changes that I needed to introduce to 
rbpf for my project. I've been working on a heavily-modified fork for a while and added an ARMv7 JIT to it. Given the scope of those changes and the fact that they are coupled with my specific requirements, I don't think I will be trying to make them mergeable with the upstream rbpf (although if you want to have a look it's [here](https://github.com/SzymonKubica/rbpf-for-microcontrollers/tree/7cc7b52b479102c4c41ab3abe411ba3de7633414))

I think compared to those changes, adding no_std support is potentially useful so I will continue working on this PR

SzymonKubica commented on src/without_std.rs:
```diff
@@ -0,0 +1,35 @@
+    pub use core::u32;
+    pub use core::u64;
+
+    // Adapt this with whatever printing functionality your host OS requires.
```
Here the idea is that in no_std there is no println! macro available by default. Different environments provide their own versions of that macro. For instance, for my use-case, I'm using Rust with RIOT OS and they provide a separate print
macro that allows for printing things to the shell available over the serial port. The idea here is that the print functionality should be overridden with the one that users have access to in their no_std environment. 

This is not a copy-paste, however I agree that it is a somewhat hacky way of allowing to use println in no_std. I will try
to come up with something better. Here are all of the current usages of println macro:
![image](https://github.com/qmonnet/rbpf/assets/50265654/412c8772-c750-4ae0-88ac-e5b2c06543ee)

They can be divided into three categories:
- print statements used in example files
- printing used as logging (e.g. in the disassembler)
- printing used inside of a helper function (`bpf_trace_printf`)

What I suggest that we do:
- leave example files as they are but ensure that they only run with std enabled
- for the logging use-cases, I suggest that we replace those calls with usages of the logging macros from the `log` crate
- for the printing helper functions, I think we can feature-gate those to be only available with std.

That way we don't have to have this hacky override that needs to be set by the users. 
The reason for using `log` is that in no_std it is able to correctly identify the printing macro that is exposed by the platform that we are running on (even if it has no standard println! macro) For instance in RIOT calling the log macros correctly prints to the shell available over the serial port.

qmonnet replied:
I suppose it's OK to use the `log` crate, no objection. The rest of your proposal sounds fine as well.

SzymonKubica commented on src/ebpf.rs:
```diff
@@ -15,6 +15,7 @@
 //! the list of the operation codes: <https://github.com/iovisor/bpf-docs/blob/master/eBPF.md>
 
 use byteorder::{ByteOrder, LittleEndian};
+use crate::stdlib::vec::Vec;
```
In order to use Vec in no_std we need to explicitly import it. However now that I look at this setup with different modules:
with_alloc, with_std, without_str it looks a bit too cluttered. I had a look at other crates and how they are doing no_std/std conditional switching and found an example from serde [here](https://docs.rs/serde/latest/src/serde/lib.rs.html#95:~:text=///%20A%20facade%20around,num%3A%3ASaturating%3B%0A%7D). It effectively does the same thing as the three of my with_* modules do but in a much cleaner way. I think I will reimplement the setup that there are using

qmonnet replied:
OK I was just not sure why we suddenly needed to import it whereas that was not the case so far.

Good thing if you find a cleaner way :+1:

SzymonKubica commented:
One note regarding your comment about asm_parser and disassembler being disabled:
I have looked into those and I don't see anything which would prevent them from running under no_std. I will update those so that they don't have to be disabled. It was probably an oversight on my part as I was primarily interested in using the VM.

SzymonKubica commented:
Ok I did some restructuring and now the import switching is done the same way serde does it.
I also changed some dependencies and now all modules build with both std/no_std.\
I will look into the jit feature and see if I can make it work with windows (it could be that I can make the  JIT enabled in no_std, in which case I could simply revert to using the windows feature flag)

SzymonKubica commented on examples/rbpf_plugin.rs:
```diff
@@ -42,7 +42,7 @@ fn main() {
                     return;
                 }
-                #[cfg(not(windows))] {
+                #[cfg(jit)] {
```
I think I might have a simpler solution. If we consider the possible cases, we are either:
- running on linux with std (in which case we want the jit enabled by default)
- running on windows (jit currently disabled)
- running with no_std (we want jit to be disabled as it requires the posix functions from libc for the jitted program buffer)
We can capture the above cases by enabling the jit only when both `feature = "std"` and `not(windows)` and disabling it when we are either on windows or no_std (or both), I'm currently doing the conditional switching as follows:
```
    #[cfg(any(not(feature = "std"), windows))] {
        assert_eq!(vm.execute_program(mem, mbuff).unwrap(), 0x2211);
    }
    #[cfg(all(feature = "std", not(windows)))] {
        // Here again we JIT-compile the program.
        vm.jit_compile().unwrap();

        // Here we must provide both a reference to the packet data, and to the
        // metadata buffer we use.
        unsafe {
            assert_eq!(vm.execute_program_jit(mem, mbuff).unwrap(), 0x2211);
        }
    }

```

qmonnet replied:
Yes, it works as well. It's more verbose in the conditions and less explicit that the `jit` feature, but it certainly boils down to the same. I'm OK with that, we can always look at feature aliases as a follow-up if we want.

qmonnet commented on README.md:
```diff
@@ -688,8 +688,8 @@ details.
 * [_Building a simple JIT in
-  Rust_](http://www.jonathanturner.org/2015/12/building-a-simple-jit-in-rust.html),
-  by Jonathan Turner (2015)
+  Rust_](https://www.sophiajt.com/building-a-simple-jit-in-rust),
+  by Sophia Turner (2015)
```
This looks like it was picked by mistake while rebasing/merging, it's not related to the PR.

qmonnet commented on src/ebpf.rs:
```diff
@@ -18,7 +18,7 @@ use byteorder::{ByteOrder, LittleEndian};
 use crate::stdlib::vec::Vec;
 
 /// Maximum number of instructions in an eBPF program.
-pub const PROG_MAX_INSNS: usize = 4096;
+pub const PROG_MAX_INSNS: usize = 1000000;
```
This looks like it was picked by mistake while rebasing/merging, it's not related to the PR.

SzymonKubica replied:
Yes this was a merge, I will checkout the latest upstream and apply the clean changes there.

qmonnet commented on examples/rbpf_plugin.rs:
```diff
@@ -5,8 +5,7 @@
 // Helper function used by https://github.com/Alan-Jowett/bpf_conformance/blob/main/tests/call_unwind_fail.data
-fn _unwind(a: u64, _b: u64, _c: u64, _d: u64, _e: u64) -> u64
-{
+fn _unwind(a: u64, _b: u64, _c: u64, _d: u64, _e: u64) -> u64 {
```
You've got a number of changes that are unrelated to the rest of the PR in this file. Let's leave formatting fixes out of this, please.

SzymonKubica replied:
Sorry about the formatting fixes, my autoformatter must have kicked in, I will make sure that the final changes don't touch the code formatting

qmonnet commented:
Can you please [sign-off](https://wiki.linuxfoundation.org/dco) your commits?

SzymonKubica commented:
Re: signing off commits
Yes I will do that from now on, I think some of my previous commits were already signed-off, but I noticed that the build doesn't go through because the old ones weren't signed-off. I will apply the clean changes by signing off commits and force push the update

SzymonKubica commented:
Ok @qmonnet I think the PR is in much better state now. Following your advice I have split my changes into two logical commits:
- adding the `"std"` feature and modifying the feature-gating for the jit to require that both `not(windows)` and `feature = "std"` are set.
- adding the proper reexports for `no_std`
I updated the description of the PR, now everything builds correctly and all available tests pass in both std and no_std. 

I have also tested that it builds correctly on an actual no_std target: STM32-f439zi running RIOT. I managed to flash this example program: 
```
#![no_std]

use riot_wrappers::riot_main;
use riot_wrappers::println;
use riot_wrappers::riot_sys::malloc;

extern crate rust_riotmodules;
extern crate rbpf;
extern crate alloc;

riot_main!(main);

fn main() {
    let prog = &[
         0x79, 0x11, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, // Load mem from mbuff into R1.
         0x69, 0x10, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, // ldhx r1[2], r0
         0x95, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00  // exit
     ];
     let mem = &mut [
         0xaa, 0xbb, 0x11, 0x22, 0xcc, 0xdd
     ];

     // Just for the example we create our metadata buffer from scratch, and we store the
     // pointers to packet data start and end in it.
     let mut mbuff = [0u8; 32];
     unsafe {
         let mut data     = mbuff.as_ptr().offset(8)  as *mut u64;
         let mut data_end = mbuff.as_ptr().offset(24) as *mut u64;
         *data     = mem.as_ptr() as u64;
         *data_end = mem.as_ptr() as u64 + mem.len() as u64;
     }

     // Instantiate a VM.
     let mut vm = rbpf::EbpfVmMbuff::new(Some(prog)).unwrap();

     // Provide both a reference to the packet data, and to the metadata buffer.
     let res = vm.execute_program(mem, &mut mbuff).unwrap();
     assert_eq!(res, 0x2211);
     println!("res: 0x{:x}", res);
}
```
It looks like everything works in `no_std`. One note: I ended up having to disable the `assembler` and `asm_parser` as they depend on the `combine` crate which doesn't have the `no_std` mode. 

I can quickly publish a repo with the experimental setup that I used to test the new version of rbpf running without `std` in RIOT if you would find this helpful.

qmonnet commented on examples/disassemble.rs:
```diff
@@ -1,37 +1,40 @@
 // SPDX-License-Identifier: (Apache-2.0 OR MIT)
 // Copyright 2017 6WIND S.A. <quentin.monnet@6wind.com>
 
+
```
No need to add a blank line here.

qmonnet commented on examples/disassemble.rs:
```diff
@@ -1,37 +1,40 @@
-        0x95, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
-    ];
-    disassembler::disassemble(prog);
+    #[cfg(feature = "std")] {
```
What happens if we run the example without `std`, nothing? Should there be a message instead, or a way to skip compilation entirely in that case?

SzymonKubica replied:
Good point, I managed to disable the entire example when testing without `std`

qmonnet commented on examples/load_elf.rs:
```diff
@@ -49,85 +49,88 @@ use rbpf::helpers;
-        println!("Packet #2, program returned: {:?} ({:#x})", res, res);
-        assert_eq!(res, 0);
+    // This example relies on the bpf_trace_printf which is only available on std
+    #[cfg(feature = "std")] {
```
Do you know if there's a way to mark the whole file as dependent on the feature, and skip this example entirely if `std` is not selected?

SzymonKubica replied:
That's probably the best way of doing this, I will try to see if it is possible, so far I've tried using: 
```rust
#![cfg(feature = "std")]
```
At the top level of the file but it seems like the example still gets included and the tests complain about it not having the  main function.

SzymonKubica replied:
Turns out `Cargo.toml` lets you specify which features the examples depend on.

qmonnet commented on examples/uptime.rs:
```diff
@@ -11,65 +11,67 @@ use rbpf::helpers;
-        0x0c, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // add32 r0, r1
-        0x95, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00  // exit and return r0
-    ];
+    #[cfg(feature = "std")] {
```
Same remark here, I'd rather skip building the example entirely if we can't run it rather than indenting all the code here and having nothing happening when `std` is turned off.

qmonnet commented on src/disassembler.rs:
```diff
@@ -4,7 +4,9 @@
 //! Functions in this module are used to handle eBPF programs with a higher level representation,
 //! for example to disassemble the code into a human-readable format.
 
+use crate::lib::*;
 use ebpf;
+use log::warn;
```
How do you sort these? I'd maybe keep local imports last?

qmonnet commented on src/helpers.rs:
```diff
@@ -250,6 +254,7 @@ pub fn strcmp (arg1: u64, arg2: u64, arg3: u64, unused4: u64, unused5: u64) -> u
 /// extern crate time;
 ///
 /// unsafe {
+///
```
Why the blank line?

qmonnet commented on src/interpreter.rs:
```diff
@@ -5,8 +5,8 @@
 
-use std::collections::HashMap;
-use std::io::{Error, ErrorKind};
+
```
What the blank line?

qmonnet commented on src/lib.rs:
```diff
@@ -27,9 +27,17 @@
     )
 )]
 
+#![cfg_attr(not(feature = "std"), no_std)]
+
```
This one is complex to read, please add a comment. Do we need the blank line? (If it applies only to the byteorder crate, probably not?)

qmonnet commented on src/lib.rs:
```diff
@@ -59,6 +66,63 @@ mod interpreter;
+#[cfg(not(feature = "std"))]
+mod no_std_error;
+
+/// A facade around all the types needed from the `std`, `core`, and `alloc`
```
“facade”: I didn't know it existed without the cedilla! :exploding_head:

qmonnet commented on src/lib.rs:
```diff
@@ -724,37 +789,39 @@ impl<'a> EbpfVmFixedMbuff<'a> {
+    ///
+    ///     let res = vm.execute_program(mem).unwrap();
+    ///     assert_eq!(res, 3);
+    ///     }
```
Should be unindented?

```suggestion
    /// }
```

qmonnet commented on src/no_std_error.rs:
```diff
@@ -0,0 +1,35 @@
+}
+
+impl Error {
+    /// New function added for compatibility with the existing code.
```
“the existing code” only makes sense in the context of the PR. After this is merged, it doesn't mean much: everything is existing code. Could you please rephrase?

qmonnet commented on src/no_std_error.rs:
```diff
@@ -0,0 +1,35 @@
+}
+
+/// The minimum set of variants to make the dummy ErrorKind work with
+/// the existing code.
```
Same comment.

qmonnet commented on tests/misc.rs:
```diff
@@ -17,9 +17,11 @@
 // use std::path::PathBuf;
 
-extern crate rbpf;
+// These tests should only be run on std as they depend on the assembler.
+#![cfg(feature = "std")]
```
Do they? It seems a lot of them don't need the assembler. Plus most of them seem already gated behind individual `std` directives?

SzymonKubica replied:
Good point, I will enable a more fine-grained control so that everything that can possibly be run runs when testing without std

qmonnet commented on tests/ubpf_verifier.rs:
```diff
@@ -16,11 +16,15 @@
 use rbpf::assembler::assemble;
 use rbpf::ebpf;
 
+
```
No need for that

qmonnet commented on tests/ubpf_verifier.rs:
```diff
@@ -16,11 +16,15 @@
 
 // These are unit tests for the eBPF “verifier”.
 
+// These tests should only be run on std as they depend on the assembler.
+#![cfg(feature = "std")]
```
Not all of them do, but OK. I'd adjust the comment to say "most" or even "many of these tests", rather than implying they all do.

SzymonKubica replied:
Yes, you are right indeed. It was a mistake on my part, I saw that compilation failed because the module was trying to import the assembler so I just disabled the whole test file. Now it is fixed, the assembler gets included conditionally and only the tests that depend on it are disabled when testing `no_std`

qmonnet reviewed (commented):
Looks much better, thanks! Some comments here and there.

qmonnet commented on Cargo.toml:
```diff
@@ -24,10 +24,13 @@ include = [
-libc = "0.2"
-time = "0.2"
-byteorder = "1.2"
+byteorder = { version = "1.2", default-features = false }
+log = {version = "0.4.21", default-features = false }
```
What does `default-features = false` do, please?

SzymonKubica replied:
This is a standard way of telling the dependencies that we are using them with `no_std`. In most cases, the way those libraries are designed is similar to how I specified in `Cargo.toml`: 
```toml
[features]
default = ["std"]
std = ["dep:time", "dep:libc", "dep:combine"]
```
And so if someone was to include `rbpf` in their project and specified: 
```toml
rbpf = { version = "1.2", default-features = false }
```
Then the `std` feature would be disabled and so the crate behaves in a way that is compatible with `no_std`.
```toml
[dependencies]
riot-wrappers = { version = "0.8", features = [ "set_panic_handler" ] }
rbpf = {path = "../../../../vms/rbpf-for-microcontrollers", default-features = false}

# While currently this example does not use any RIOT modules implemented in
# Rust, that may change; it is best practice for any RIOT application that has
# its own top-level Rust crate to include rust_riotmodules from inside
# RIOTBASE.
rust_riotmodules = { path = "../../sys/rust_riotmodules/" }
```
This is an example of how I included `rbpf` without default features in that example program with RIOT on stm32

SzymonKubica replied:
I can add a comment in the `Cargo.toml` explaining why we don't need the default features

qmonnet commented:
Please clean up the commit log from your second commit, you have multiple sign-off tags (likely because you squashed commits together and kept the descriptions of all commits).

SzymonKubica commented on src/asm_parser.rs:
```diff
@@ -593,7 +598,7 @@ exit
         assert_eq!(
             parse(
-                " 
+                "
```
Not sure what happened here, I tried bringing the space back but it seems like it gets clipped when I commit (it could be that my editor clipps dangling whitespace) or some linux CRLF / LF issue

SzymonKubica replied:
Ok I used a different editor and managed to get this diff to disappear

SzymonKubica commented:
Ok I think all of the comments have been addressed. Thank you very much for the feedback. 
Key changes introduced:
- now examples are configured inside `Cargo.toml` and the ones that need std are disabled when testing on `no_std`
- the `misc`, `ubpf_verfier`, `ubpf_vm` tests now run all available tests under `no_std` (no more coarse-grained file level feature-gating) Thank you for picking up on this issue. 
- removed whitespace and formatting issues that were introduced by mistake
- reordered the imports as you suggested
- added required explanation / clarification comments

SzymonKubica commented on src/lib.rs:
```diff
@@ -42,15 +44,37 @@ extern crate cranelift_module;
 #[cfg(feature = "cranelift")]
 mod cranelift;
+#[cfg(std)]
 pub mod disassembler;
```
Yes, I agree. I'm not sure what would be the best place for this doc. I think I could add a section similar to [this one](https://serde.rs/no-std.html) to the readme. Would this work?

SzymonKubica commented:
Hi @qmonnet, 
I added a description of the `no_std` feature to the README, it was based on how `serde` explains their implementation of `no_std` support. This was done to resolve one of your eariler comments that it isn't obvious what is available and what isn't under
`no_std`. I think this was the last comment that was left to be addressed, so feel free to have another look at the PR when you're available. Thank you for all of your feedback, it was really helpful to learn how to prepare a patch properly.

qmonnet commented on src/assembler.rs:
```diff
@@ -1,15 +1,17 @@
 // Copyright 2017 Rich Lane <lanerl@gmail.com>
 
-//! This module translates eBPF assembly language to binary.
+//! This module translates eBPF assembly language to binary. It depends on the
+//! `combine` crate, which depends on the standard library. Because of this
+//! it is only available when the `std` feature is enabled.
```
Looking at the amount of tests that you had to leave out when we use `std`, it's rather annoying to leave the assembler out. How much did you look into `combine`'s support for no_std? [Looking at its changelog](https://github.com/Marwes/combine/blob/f726b8f943f44d258686911907c0324771a2d09f/CHANGELOG.md?plain=1#L446), it should be able to work without `std` since v3 (we're using v4+). Could we deselect `std` for that crate when we deselect it for rbpf, keep building the assembler (and disassembler if possible), and keep the tests?

SzymonKubica replied:
Good point, I will try and enable it again tonight. From what I remember, simply specifying `default-features=false` in the Cargo.toml didn't cut it but given the change log that you have shared there might still be a way to use it under no_std. It could be that the alloc feature needs to be enabled. 

I do agree that quite a few tests needed to be left out and it would be good to enable those. 
As for the disassembler, this module seems to work fine with `no_std`, it is just that some of its tests require the assembler to prepare the test case which means that they need to be disabled

SzymonKubica replied:
The changelog does indeed look promising, I was probably just misconfiguring something. I don't think there is something that needs to happen when rbpf calls into `combine` that would require the standard library. From what I could tell it was some parser combinators stuff

SzymonKubica replied:
Oh, I doublechecked and unfortunately it seems like `rbpf` depends on the `EasyParser` from combine which for some reason is available on `std` only (see [here](https://docs.rs/combine/latest/combine/trait.EasyParser.html#:~:text=Available%20on%20crate%20feature%20std%20only.)

In this case there are two ways of moving forward: 
- leave `asm_parser` and assembler disabled
- change the implementation of the `asm_parser` to not depend on the `EasyParser`

From what I can tell, there is only one place where `rbpf` depends on that trait:
```rust
/// Parse a string into a list of instructions.
///
/// The instructions are not validated and may have invalid names and operand types.
pub fn parse(input: &str) -> Result<Vec<Instruction>, String> {
    match spaces()
        .with(many(instruction()).skip(eof()))
        .easy_parse(position::Stream::new(input))
    {
        Ok((insts, _)) => Ok(insts),
        Err(err) => Err(err.to_string()),
    }
}
```
According to the docs, the easy parser is supposed to produce better error messages. However in this place the error is turned directly into a string, so the code doesn't actually depend on the structue of the returned errors. I will and see if I can replace that with the normal parser. There are two possible approaches:
- replace this with the regular parser and sacrifice the nice error messagea
- instrument the code to use `EasyParser` when we have access to `std` and `Parser` otherwise

qmonnet replied:
If we end up skipping two tests only, then yes it's absolutely fine! Thanks a lot for looking into this!

SzymonKubica replied:
Yes, it turned out that the only difference between `EasyParser` and the `Parser` in combine is that they produce different error messages. Only these two test cases above depend on the contents of those errors, so I added some conditional `cfg` to change the expedted error depending on the enabled features. Aside from this all other tests behave the same under both environments.

qmonnet commented on Cargo.toml:
```diff
@@ -43,11 +48,33 @@ json = "0.11"
+
+[[example]]
+name = "rbpf_plugin"
+
```
```suggestion
```

qmonnet commented on README.md:
```diff
@@ -19,6 +19,7 @@ Rust (user-space) virtual machine for eBPF
 * [Building eBPF programs](#building-ebpf-programs)
 * [Feedback welcome!](#feedback-welcome)
 * [Questions / Answers](#questions--answers)
+* [No_std Support](#no_std-support)
```
Let's maybe have a “Build features” section (or something like this) instead, and move it higher up in the list (above “Feedback welcome!” maybe). Then we can have `no_std` as a subsection of that? We need to document the Cranelift JIT at some point (not in this PR of course), and it would be another subsection of “Build features”.

qmonnet commented on examples/load_elf.rs:
```diff
@@ -48,7 +48,6 @@ use rbpf::helpers;
 // Once the bytecode has been edited, we can load the bytecode directly from the ELF object file.
 
 fn main() {
-
```
Unrelated

qmonnet commented on tests/misc.rs:
```diff
@@ -18,8 +18,8 @@
 // use std::path::PathBuf;
 
 extern crate rbpf;
-
```
Let's keep an empty line here?

qmonnet commented on tests/ubpf_vm.rs:
```diff
@@ -9,7 +9,6 @@
 // <https://github.com/iovisor/ubpf/tree/master/tests> (hyphen had to be replaced with underscores
 // as Rust will not accept them in function names). It is strongly advised to refer to the uBPF
 // version to understand what these program do.
-//
```
Please keep the line :)

SzymonKubica replied:
Yes sorry about that, must have deleted it accidentally and didn't spot it in the diff. It was inserted back.

qmonnet reviewed (changes requested):
Some small nits

qmonnet commented:
It could be interesting to add the following change as an additional commit to your PR:

```diff
diff --git a/.github/workflows/test.yaml b/.github/workflows/test.yaml
index 60eefc28d727..a11391c430ca 100644
--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -17,10 +17,11 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
+        toolchain: [stable, beta, nightly]
+        features: [--all-features]
         include:
-          - toolchain: stable
-          - toolchain: beta
           - toolchain: nightly
+            features: --no-default-features
 
     steps:
       - name: Checkout code
@@ -33,12 +34,12 @@ jobs:
           override: true
           components: clippy
 
-      - name: Build with ${{ matrix.toolchain }}
+      - name: Build with ${{ matrix.toolchain }}, ${{ matrix.features }}
         run: |
           cargo +${{ matrix.toolchain }} build \
-              --release --all-features --all-targets
+              --release ${{ matrix.features }} --all-targets
 
-      - name: Test with ${{ matrix.toolchain }}
+      - name: Test with ${{ matrix.toolchain }}, ${{ matrix.features }}
         run: |
           if [[ "${{ matrix.toolchain }}" == 'nightly' ]]; then
               export RUSTDOCFLAGS='-Zsanitizer=address'
@@ -47,7 +48,7 @@ jobs:
           fi
           cargo +${{ matrix.toolchain }} test \
               --target=x86_64-unknown-linux-gnu \
-              --all-features
+              ${{ matrix.features }}
 
       - name: Lint with ${{ matrix.toolchain }}
         run: |
```

I've not tested it, but if I didn't make any mistake it should add a workflow job to build and run the (available) tests without the `std` feature.

SzymonKubica commented:
> It could be interesting to add the following change as an additional commit to your PR:
> 
> ```diff
> diff --git a/.github/workflows/test.yaml b/.github/workflows/test.yaml
> index 60eefc28d727..a11391c430ca 100644
> --- a/.github/workflows/test.yaml
> +++ b/.github/workflows/test.yaml
> @@ -17,10 +17,11 @@ jobs:
>      strategy:
>        fail-fast: false
>        matrix:
> +        toolchain: [stable, beta, nightly]
> +        features: [--all-features]
>          include:
> -          - toolchain: stable
> -          - toolchain: beta
>            - toolchain: nightly
> +            features: --no-default-features
>  
>      steps:
>        - name: Checkout code
> @@ -33,12 +34,12 @@ jobs:
>            override: true
>            components: clippy
>  
> -      - name: Build with ${{ matrix.toolchain }}
> +      - name: Build with ${{ matrix.toolchain }}, ${{ matrix.features }}
>          run: |
>            cargo +${{ matrix.toolchain }} build \
> -              --release --all-features --all-targets
> +              --release ${{ matrix.features }} --all-targets
>  
> -      - name: Test with ${{ matrix.toolchain }}
> +      - name: Test with ${{ matrix.toolchain }}, ${{ matrix.features }}
>          run: |
>            if [[ "${{ matrix.toolchain }}" == 'nightly' ]]; then
>                export RUSTDOCFLAGS='-Zsanitizer=address'
> @@ -47,7 +48,7 @@ jobs:
>            fi
>            cargo +${{ matrix.toolchain }} test \
>                --target=x86_64-unknown-linux-gnu \
> -              --all-features
> +              ${{ matrix.features }}
>  
>        - name: Lint with ${{ matrix.toolchain }}
>          run: |
> ```
> 
> I've not tested it, but if I didn't make any mistake it should add a workflow job to build and run the (available) tests without the `std` feature.

Thank you very much for proposing this change, It would have taken me a lot longer to figure out how the ci works and how to enable the `no_std` in the tests workflow. I realise that the last pipeline has failed but that was because I was missing the toolchain section in the `include`. Now I think it should work correctly, the test have failed because there was not toolchain specified in the last `no_std` test run

qmonnet commented:
The tests pass correctly this time :tada:, but we get warnings about `cargo-clippy`. I'm not sure why, maybe using `--no-default-features` disables something clippy-related. I'll take a look later.

SzymonKubica commented:
> The tests pass correctly this time 🎉, but we get warnings about `cargo-clippy`. I'm not sure why, maybe using `--no-default-features` disables something clippy-related. I'll take a look later.

I think the reason this started producing warnings was that running `cargo clippy` implicitly adds a `cargo-clippy` feature that isn't included in the `Cargo.toml`. It looks like `rbpf` uses a legacy way of specifying clippy lint attributes (possible explanation [here](https://blog.rust-lang.org/2024/02/28/Clippy-deprecating-feature-cargo-clippy.html)). Because of this, when running with --no-default-feature this `cargo-clippy` feature is missing and we get warnings. 

I also looked through the test logs and got rid of all other warnings inside of misc.rs and rbpf_plugin.rs. All of them were mostly unused imports or functions that weren't used when running in `no_std`.

After these changes all builds were passing and tests didn't introduce any new warnings. However,  I needed to squash the fix commits to make a single one that deals with all of the warnings. Could you please approve this workflow so that we can see that indeed everything got fixed? Thank you for your time, it's starting to get very exciting as we are getting closer having everything integrated :tada:

SzymonKubica commented:
Ok @qmonnet I think all checks were successful and after running tests with `--no-default-features` there were no warnings emitted :tada: :rocket: . Thank you for helping me enable tests for the `no_std` feature. Do let me know if anything else requires attention in the PR and I will adjust it.

qmonnet commented:
Thanks a lot! Sorry I didn't have the time to review last night, I'll do it soon.

SzymonKubica commented:
> Thanks a lot! Sorry I didn't have the time to review last night, I'll do it soon.

No worries at all, take your time. No need to rush, I realise it is quite a big change in terms of files affected (not that big functionality-wise). Thank you for your time and guidance with the PR.

qmonnet commented on src/lib.rs:
```diff
@@ -59,6 +67,61 @@ mod interpreter;
+    pub use alloc::string::{String, ToString};
+    #[cfg(feature = "std")]
+    pub use std::string::{String, ToString};
+
+
```
Nit: Double empty line, I think I've been trying to avoid these in the rest of the repo (although a few sneaked in)

SzymonKubica replied:
Sure I will remove that, thanks for picking up on these!

qmonnet commented on src/no_std_error.rs:
```diff
@@ -0,0 +1,42 @@
+    }
+}
+
+/// The current version of `rbpf` only uses the [`Other`](ErrorKind::Other) variant
+/// from the [std::io::ErrorKind] enum. If a dependency on other variants were
+/// introduced in the future, this enum needs to be updated accordingly to maintain
+/// compatibility with the real `ErrorKind`. The reason all available variants
+/// aren't included in the first place is that [std::io::ErrorKind] exposes
+/// 40 variants, and not all of them are meaningful under `no_std`.
```
Thanks for documenting this! :+1:

qmonnet commented on tests/assembler.rs:
```diff
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: (Apache-2.0 OR MIT)
 // Copyright 2017 Rich Lane <lanerl@gmail.com>
 
+// These tests should only be run on std as they depend on the assembler.
+#![cfg(feature = "std")]
```
I'm confused here as well, can't we enable some of these tests now?

SzymonKubica replied:
I think those were reenabled in one of the subsequent commits. I decided treat enabling no_std in the assembler and disassembler (and enabling their corresponding tests) as a separate feature as it required changing the way we use `combine` library. I can squash the two commits together so that it is no longer confusing

SzymonKubica replied:
Oh I doublechecked, sorry about this. Those can indeed be enabled without issues

qmonnet commented on README.md:
````diff
@@ -559,6 +560,26 @@ program.add(Source::Imm, Arch::X64).set_dst(1).set_imm(0x605).push()
+The `rbpf` crate has a Cargo feature named "std" that is enabled by default. To
+use `rbpf` in `no_std` environments this feature needs to be disabled. To do
+this, you need to modify your dependency on `rbpf` in Cargo.toml to disable the
+enabled-by-default features.
+```toml
+[dependencies]
+rbpf = { version = "1.0", default-features = false }
+```
+Note that when using this crate in `no_std` environments, the `assembler`,
````
Nit: Blank lines around the literal block, please

````suggestion
enabled-by-default features.

```toml
[dependencies]
rbpf = { version = "1.0", default-features = false }
```

Note that when using this crate in `no_std` environments, the `assembler`,
````

qmonnet commented on src/lib.rs:
```diff
@@ -43,11 +51,11 @@ extern crate cranelift_module;
+#[cfg(feature = "std")]
 mod asm_parser;
+#[cfg(feature = "std")]
 pub mod assembler;
```
Ah but I see you re-enable it in one of the next commits, sorry.

SzymonKubica replied:
Oh yes, sorry about this. I think it was originally disabled but then I tried harder to support it and it was probably included in one of the subsequent logical changes

qmonnet reviewed (changes requested):
Thanks a lot for your work!

I've got some comments, and I think I'm a bit confused as to when the assembler is supported without `std`. I probably need to take some time to play with your code and check why rbpf complains if I try to enable the tests; in the meantime, I commented at the locations where it is not clear to me.

SzymonKubica commented on Cargo.toml:
```diff
@@ -44,11 +49,31 @@ hex = "0.4.3"
+[[example]]
+name = "uptime"
+required-features = ["std"]
+
+[[example]]
+name = "to_json"
+
+[[example]]
+name = "rbpf_plugin"
```
Yes, unfortunately we do. The way it works is that once you explicitly specify one example in `Cargo.toml` all other examples need to be specified as well. I remember seeing some warning when I didn't include these. I'm going to find it and put it here.

SzymonKubica replied:
This is the error message that pops up when examples aren't explicitly specified: 
```
warning: An explicit [[example]] section is specified in Cargo.toml which currently
disables Cargo from automatically inferring other example targets.
This inference behavior will change in the Rust 2018 edition and the following
files will be included as a example target:

* /home/szymon/Projects/ebpf-on-microcontrollers/vms/rbpf-for-microcontrollers/examples/to_json.rs
* /home/szymon/Projects/ebpf-on-microcontrollers/vms/rbpf-for-microcontrollers/examples/rbpf_plugin.rs

This is likely to break cargo build or cargo test as these files may not be
ready to be compiled as a example target today. You can future-proof yourself
and disable this warning by adding `autoexamples = false` to your [package]
section. You may also move the files to a location where Cargo would not
automatically infer them to be a target, such as in subfolders.

```

SzymonKubica commented on README.md:
````diff
@@ -497,19 +504,21 @@ Conversely, a disassembler is also available to dump instruction names from
 ```rust
-extern crate rbpf;
-use rbpf::disassembler::disassemble;
+#[cfg(feature = "std")] {
+    extern crate rbpf;
+    use rbpf::disassembler::disassemble;
````
Thank you for picking up on this one, yes I think this one should be good. I will confirm locally  and update the PR

SzymonKubica commented on src/asm_parser.rs:
```diff
@@ -4,7 +4,9 @@
 // file of the crate. Do not expect to find those comments in the documentation of the crate.
 
-//! This module parses eBPF assembly language source code.
+//! This module parses eBPF assembly language source code. It depends on the
+//! `combine` crate, which depends on the standard library. Because of this
+//! it is only available when the `std` feature is enabled.
```
Yes I will amend this

SzymonKubica commented on src/assembler.rs:
```diff
@@ -1,15 +1,17 @@
 // Copyright 2017 Rich Lane <lanerl@gmail.com>
 
-//! This module translates eBPF assembly language to binary.
+//! This module translates eBPF assembly language to binary. It depends on the
+//! `combine` crate, which depends on the standard library. Because of this
+//! it is only available when the `std` feature is enabled.
```
Yes this got out of date, will update now

SzymonKubica commented on src/cranelift.rs:
```diff
@@ -28,6 +21,7 @@ use crate::ebpf::{
 };
 
 use super::Error;
+use crate::lib::*;
```
Totally agree, that way all crate imports are in one place.

SzymonKubica commented on src/disassembler.rs:
````diff
@@ -383,6 +386,7 @@ pub fn to_insn_vec(prog: &[u8]) -> Vec<HLInsn> {
 /// neg64 r2
 /// exit
 /// ```
+#[cfg(feature = "std")]
 pub fn disassemble(prog: &[u8]) {
````
I have updated it so that it calls log::info! when under no_std. I think it basically replicates the behaviour without disabling it. The idea is that in `no_std` one can define a custom logger that calls into the OS-specific printing functionality that can then print to a log file or the serial port like so:
```
impl Log for RiotLogger {
    fn enabled(&self, metadata: &Metadata<'_>) -> bool {
        metadata.level() <= self.level
    }

    fn log(&self, record: &Record<'_>) {
        if self.enabled(record.metadata()) {
            println!(
                "{}:{} -- {}",
                record.level(),
                record.target(),
                record.args()
            );
        }
    }

    fn flush(&self) {}
}
```
The example above is the logger that I used for RIOT applications.

qmonnet replied:
> The example above is the logger that I used for RIOT applications.

Just to double-check and be 100% clear here, this is some code that _you_ contributed to RIOT, not something in RIOT from which you took inspiration, right? I'm asking because I see RIOT is LGPL, so we can't reproduce its code in rbpf if you weren't the one to contribute it in the first place (although reading again, you mention RIOT _applications_ which is maybe something different).

SzymonKubica replied:
Yes that's right, this is just some custom application-specific code that I have in an app built on top of riot. Additionally, this snippet of code was just to illustrate to you how one might go about implementing the logger behaviour in their system. It's not actually contributed anywhere in rbpf. There is nothing in this PR that was taken / inspired by RIOT, I was just trying to motivate why using the log crate is a good idea here to allow users to specify where things should get logged to

qmonnet replied:
OK perfect, thank you for confirming :slightly_smiling_face:

SzymonKubica commented:
> Thanks a lot for your work!
> 
> I've got some comments, and I think I'm a bit confused as to when the assembler is supported without `std`. I probably need to take some time to play with your code and check why rbpf complains if I try to enable the tests; in the meantime, I commented at the locations where it is not clear to me.
> 
> The required changes are mostly minor things, we're getting closer!
> 
> Regarding the formatting of your patches, I'd recommend the following:
> 
> * Squash the fixes from the last commit into the relevant earlier commits
> * Rework your commit descriptions, in particular for the second commit. What's interesting in this description is not so much to describe the changes (and not the history of the different changes and fixes incorporated during the review); instead, it should provide the motivation for the change, and explain what the commit addresses, what difficulties arose and how the commit works around them, etc. Everything related to the context of the change.

Ok thank you for the review and your suggestions. I will rework the commit messages and squash the subsequent fixes.
I propose the following logical changes:
- add the `std` feature and updating the windows cfg flags
- add lib module and conditional imports (the **Configure combine crate and asm_parser to work under no_std.** commit will be squashed into this one)
- update readme with feature description
- add workflow job

The last commit fixing pipeline warnings will be split and squashed into the above commits where the changes belongj appropriately

SzymonKubica commented:
Ok I think everything has been cleaned up now. Thank you very much for all of your feedback and time spent reviewing my changes. Now all imports / dependency changes are in the second commit. I managed to find some additional assembler tests that could be reenabled (tests/misc.rs). I think it helps a lot to see all changes in a single logical commit

SzymonKubica commented:
@qmonnet could you please approve the CI workflow run? Don't worry about reviewing the changes today, no need to rush. I just wanted to make sure that rearranging the changes didn't produce any new pipeline warnings. Enjoy your weekend and feel free to look into the PR when you are free.

SzymonKubica commented on README.md:
````diff
@@ -555,6 +556,28 @@ program.add(Source::Imm, Arch::X64).set_dst(1).set_imm(0x605).push()
+rbpf = { version = "1.0", default-features = false }
+```
+
+Note that when using this crate in `no_std` environments, the `assembler`,
+`asm_parser` and `jit` modules aren't available. This is because the assembler
+depends on the `combine` crate providing parser combinators. This crate doesn't
+provide `no_std` support for the features that `rbpf` depends on.
````
No worries at all, I realise that maintaining an open source project alongside your main job is quite challenging. Thank you for all of your feedback in getting this PR into an acceptable state. I will quickly amend that commit to reflect the current state of the system. Thank you for picking up on it, I really appreciate the level of detail that you pay attention to.

SzymonKubica commented:
> Apologies once again for the delay. I do realise this is not a great way to keep you motivated for this PR, and I'm sorry for the poor contribution experience in terms of delays. Thanks a lot for bearing with me.
> 
> And thanks for this work, this is in a very nice shape now! I think there's just a small fix to do in the README, but everything else looks good as far as I can tell. Well done, and nice work on the commit description for your second commit! ❤️

Ok @qmonnet, thank you for pointing out the issue it the README, it is fixed now. Thank you for picking up on it.

qmonnet reviewed (approved):
Awesome, thanks a lot for this work!

SzymonKubica commented:
Perfect, thank you for approving the PR :rocket: 
I really appreciate your help along the way. It was a great learning experience and I think it is going to be very useful for my future open-source contributions. Thanks for your time and being open to this contribution.

SzymonKubica commented:
> Congrats on getting your µBPF paper into the ACM workshop! 🎉

Thank you!
